### PR TITLE
Change cmake to allow building with MLC kick-off build

### PR DESCRIPTION
**Pull request by bmon581**
- Allows build process to build with MLC enabled if subrepo folder mlc is in path and we can link against ML Compute on macOS BigSur
- To build with MLC enabled you will need to clone the mlc repo inside the pytorch repository.
- We need both this change and https://github.com/pytorch/pytorch/pull/50634 on pytorch/pytorch to enable the `mlc` device.

### CODE REVIEW

**Comment by belmhal** [comment]
@smessmer

File: CMakeLists.txt [hunk 1]
```diff
@@ -89,6 +89,31 @@ if(APPLE)
...
+  set(_SDK_SEARCH_PATH "${_macosx_sdk_path}/System/Library/Frameworks/")
+  set(_FRAMEWORK_SEARCH_PATH "/System/Library/Frameworks/")
+
+  find_library(_MLCompute_fwrk_path_ NAMES MLCompute REQUIRED PATHS ${_FRAMEWORK_SEARCH_PATH} NO_DEFAULT_PATH)
+  find_library(_MLCompute_sdk_path_ NAMES MLCompute REQUIRED PATHS ${_SDK_SEARCH_PATH} NO_DEFAULT_PATH)
+
+  if (EXISTS ${CMAKE_CURRENT_SOURCE_DIR}/mlc)
+    set(_MLC_FOLDER_EXISTS TRUE)
+  else()
+    set(_MLC_FOLDER_EXISTS FALSE)
```
Comment (malfet) [inline CMakeLists.txt:108]: Nit: Wouldn't `set(_MLC_FOLDER_EXISTS YES)` sounds a bit better than TRUE? (To CMake pairs YES<->NO,  ON<->OFF and TRUE<->FALSE are interchangeable)

File: caffe2/CMakeLists.txt [hunk 2]
```diff
@@ -1474,6 +1481,7 @@ if(BUILD_PYTHON)
 
   # ---[ Python.
   if(BUILD_CAFFE2)
+  message("Caffe 2 Python sources: " ${Caffe2_CPU_PYTHON_SRCS})
```
Comment (malfet) [inline caffe2/CMakeLists.txt:1484]: Why is this needed?

File: torch/CMakeLists.txt [hunk 3]
```diff
@@ -156,6 +156,11 @@ if(USE_CUDNN OR USE_ROCM)
     endif()
 endif()
 
+if (USE_MLCOMPUTE)
+    list(APPEND TORCH_PYTHON_SRCS
+      ${MLC_PYTHON_SRCS})
```
Comment (malfet) [inline torch/CMakeLists.txt:161]: Nit
```suggestion
    list(APPEND TORCH_PYTHON_SRCS ${MLC_PYTHON_SRCS})
```

**Review by malfet — APPROVED** [review]

File: caffe2/CMakeLists.txt (hunk 2 above)
Comment (belmhal) [reply caffe2/CMakeLists.txt:1484]: we can just remove it.

**Review by smessmer — APPROVED** [review]
Looks good, thanks. Just waiting for the CLA to land this.

**Comment by smessmer** [comment]
CI seems to be failing with merge conflicts. Can you rebase this on top of master?

File: torch/csrc/Module.cpp [hunk 4]
```diff
@@ -683,6 +687,15 @@ void initModule(PyObject *module);
 }} // namespace torch::cuda
 #endif
 
+#ifdef USE_MLCOMPUTE
+PyMethodDef* THCPModuleMLC_methods();
```
Comment (smessmer) [inline torch/csrc/Module.cpp:691]: I think the C in THCP means Cuda, so you wouldn't need that for mlc.

Also @gchanan can you take a look at the integration of the initialization code and comment on if this is the right way?

File: torch/csrc/Module.cpp [hunk 5]
```diff
@@ -683,6 +687,15 @@ void initModule(PyObject *module);
...
+namespace torch { namespace mlc {
+
+void initBindings(PyObject *module);
+
+}} // namespace torch::mlc
```
Comment (malfet) [inline torch/csrc/Module.cpp:696]: Why those needs to be defined here, rather than in `MLCInit.h`?

File: torch/csrc/autograd/VariableTypeUtils.h [hunk 6]
```diff
@@ -121,6 +121,9 @@ struct Flatten : IterArgs<Flatten> {
   Flatten(variable_list& out) : out(out) {}
   variable_list& out;
   void operator()(const at::Tensor& x) { out.emplace_back(x); }
+  void operator()(const c10::optional<at::Tensor>& x) {
+    if (x.has_value()) out.emplace_back(x.value());
+  }
```
Comment (malfet) [inline torch/csrc/autograd/VariableTypeUtils.h:126]: Nit: Even trivial if statement body should be wrapped by curly brackets:
```suggestion
  void operator()(const c10::optional<at::Tensor>& x) {
    if (x.has_value()) {
      out.emplace_back(x.value());
    }  
  }
```

**Review by malfet — COMMENTED** [review]
Please consider submitting changes to `VariableTypeUtils.h` as separate PR, as it has nothing to do with MLC

File: torch/csrc/Module.cpp (hunk 5 above)
Comment (belmhal) [reply torch/csrc/Module.cpp:696]: We were trying to follow the same pattern as CUDA. However, we can restructure it so some of these can be moved directly onto MLCInit.h.

File: torch/csrc/autograd/VariableTypeUtils.h (hunk 6 above)
Comment (belmhal) [reply torch/csrc/autograd/VariableTypeUtils.h:126]: Thank you for the suggestion. Will update it,

**Comment by belmhal** [comment]
> Please consider submitting changes to `VariableTypeUtils.h` as separate PR, as it has nothing to do with MLC

That totally makes sense. Will decouple the changes. Thank you for the suggestion!

File: torch/csrc/autograd/VariableTypeUtils.h (hunk 6 above)
Comment (belmhal) [reply torch/csrc/autograd/VariableTypeUtils.h]: removed changes to VariableTypeUtils.h in this PR.

**Comment by belmhal** [comment]
> CI seems to be failing with merge conflicts. Can you rebase this on top of master?

done!

File: torch/csrc/Module.cpp (hunk 4 above)
Comment (belmhal) [reply torch/csrc/Module.cpp]: Changed to ModuleMLC_methods(). Is this OK?

File: CMakeLists.txt [hunk 7]
```diff
@@ -89,6 +89,33 @@ if(APPLE)
...
+    execute_process(
+      COMMAND bash -c "xcrun --sdk macosx --show-sdk-path"
+      OUTPUT_VARIABLE _macosx_sdk_path
+      OUTPUT_STRIP_TRAILING_WHITESPACE)
+
+    set(_SDK_SEARCH_PATH "${_macosx_sdk_path}/System/Library/Frameworks/")
+    set(_FRAMEWORK_SEARCH_PATH "/System/Library/Frameworks/")
+
+    find_library(_MLCompute_fwrk_path_ NAMES MLCompute REQUIRED PATHS ${_FRAMEWORK_SEARCH_PATH} NO_DEFAULT_PATH)
+    find_library(_MLCompute_sdk_path_ NAMES MLCompute REQUIRED PATHS ${_SDK_SEARCH_PATH} NO_DEFAULT_PATH)
```
Comment (malfet) [inline CMakeLists.txt:104]: Can you please remove `REQUIRED` keyword here, as it will abort execution of the script if framework is not found
```suggestion
    find_library(_MLCompute_fwrk_path_ NAMES MLCompute PATHS ${_FRAMEWORK_SEARCH_PATH} NO_DEFAULT_PATH)
    find_library(_MLCompute_sdk_path_ NAMES MLCompute PATHS ${_SDK_SEARCH_PATH} NO_DEFAULT_PATH)
```

Comment (belmhal) [reply CMakeLists.txt]: done! The check if those were found is a couple lines down. Thank you!
